Add Google Analytics

edipo2s opened the pull request.

beheh commented on template.html:
```diff
@@ -27,6 +27,7 @@
 	</head>
 	<body>
 		<div id="root"></div>
+		<script src='https://www.google-analytics.com/analytics.js'></script>
```
```suggestion
		<script src="https://www.google-analytics.com/analytics.js"></script>
```

beheh commented on src/viewer/Entity.tsx:
```diff
@@ -85,6 +85,7 @@ class Entity extends React.Component<Props & CardsProps & PortalProps, State> {
 		}
 		const card = this.props.cards.getByDbfId(this.props.dbfId);
 
+		let gaEventTimeout: NodeJS.Timer;
```
This is always a bit confusing, in Node.js `setTimeout` returns `NodeJS.Timer`, but as this is a browser, it will return `number`.

beheh commented on src/viewer/Entity.tsx:
```diff
@@ -142,14 +143,20 @@ class Entity extends React.Component<Props & CardsProps & PortalProps, State> {
 						y,
 					});
+
+					gaEventTimeout = setTimeout(() => {
```
You don't need to duplicate this here. Let's just reuse `this.state.showStatistics` - we could rename it to `isMeaningfulHover` to be more clear. It's also used to fire off an HTTP request to HSReplay.net after the timeout that's set further up in this file. Then you would just do the following in `componentDidUpdate`:

```typescript
if (this.state.isHovering && !prevState.isMeaningfulHover && this.state.isMeaningfulHover) {
  ga("send", "event", "Hover", this.props.dbfId);
}
```

edipo2s replied:
hum, yeah, reuse that statistics timeout is much better!

beheh commented on src/viewer/mobile/Panel.tsx:
```diff
@@ -264,6 +264,7 @@ class Panel extends React.Component<Props & TwitchExtProps, State> {
 
 	private onCopy = () => {
 		this.setState({ copied: true });
+		ga("send", "event", "Click", "Copy Deck", "Mobile");
```
You will also need to initialize GA in `mobile/index.tsx`. If you follow my recommendation above you can just include the `analytics.ts` file there.

beheh replied:
```suggestion
		ga("send", "event", "Deck", "Copy", "Mobile");
```

beheh commented on src/viewer/overlay/index.tsx:
```diff
@@ -11,6 +11,9 @@ import {
 } from "../../utils/twitch";
 import Root from "./Root";
 
+ga.create("UA-81509238-9");
```
Let's stick with the official template.

```suggestion
ga("create", "UA-81509238-9", "auto");
```

beheh commented on src/viewer/overlay/index.tsx:
```diff
@@ -11,6 +11,9 @@ import {
 import Root from "./Root";
 
+ga.create("UA-81509238-9");
+ga.getAll().forEach(tracker => tracker.set("anonymizeIp", true));
```
Also here, although we still need to track the page view, so let's add Google's official recommendation there in the second line:

```suggestion
ga("set", "anonymizeIp", true);
ga("send", "pageview");
```

edipo2s commented:
> I see right now you're loading GA with a snippet like this:
> 
> ```
> ga.create("UA-81509238-9");
> ga.getAll().forEach(tracker => tracker.set("anonymizeIp", true));
> ```
> 
> I'm unable to find documentation for that method with a brief glance. I think we should follow Twitch's install instructions precisely by use the official Snippet from the PDF (which also happens to match Google's recommendations):
> 
> ```js
> (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
> (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
> m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
> })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
> ```
> 
> Because the official snippet is a but compact, Prettier will most likely want to break it apart. To prevent that, what about putting it into a separate `analytics.js` or `.ts` file and excluding it from prettier with a `.prettierignore`. If we then include that right at the start of `overlay/index.js`, we can then cleanly call the initialization similar to how you have done so far (although I've added two suggestions).



> I see right now you're loading GA with a snippet like this:
> 
> ```
> ga.create("UA-81509238-9");
> ga.getAll().forEach(tracker => tracker.set("anonymizeIp", true));
> ```
> 
> I'm unable to find documentation for that method with a brief glance. I think we should follow Twitch's install instructions precisely by use the official Snippet from the PDF (which also happens to match Google's recommendations):
> 
> ```js
> (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
> (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
> m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
> })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');
> ```
> 
> Because the official snippet is a but compact, Prettier will most likely want to break it apart. To prevent that, what about putting it into a separate `analytics.js` or `.ts` file and excluding it from prettier with a `.prettierignore`. If we then include that right at the start of `overlay/index.js`, we can then cleanly call the initialization similar to how you have done so far (although I've added two suggestions).

That method comes from Google analytics types from [here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/google.analytics/index.d.ts). 
But I will try to use that JavaScript code in this typeScript project again

beheh commented:
The Google Analytics types support multiple versions of GA:

> // Type definitions for Google Analytics (Classic and Universal)

As all types from DT they're also community maintained, so they might give a false sense of security here if we don't use exactly the version these typings were written for. We're in a similar spot with our TwitchExt typings, if anyone ever uses them I wouldn't be surprised if they use a snippet from the docs rather than from our typings.

edipo2s commented:
> The Google Analytics types support multiple versions of GA:
> 
> > // Type definitions for Google Analytics (Classic and Universal)
> 
> As all types from DT they're also community maintained, so they might give a false sense of security here if we don't use exactly the version these typings were written for. We're in a similar spot with our TwitchExt typings, if anyone ever uses them I wouldn't be surprised if they use a snippet from the docs rather than from our typings.

I see.. that's the reason I start to use it too, I thought would be the right way, since the twitch script are in a similar way. 😅

beheh commented on src/viewer/Entity.tsx:
```diff
@@ -63,11 +63,22 @@ class Entity extends React.Component<Props & CardsProps & PortalProps, State> {
+			this.state.isMeaningfulHover
+		) {
+			if (this.props.dbfId) {
+				ga("send", "event", "Hover", "Card", this.props.dbfId.toString());
```
I suggest casting strings this way:
```suggestion
				ga("send", "event", "Hover", "Card", "" + this.props.dbfId);
```

You can also use `String(this.props.dbfId)`, as suggested in the [Airbnb style guide](https://github.com/airbnb/javascript#coercion--strings) (a great doc to read!).

edipo2s replied:
great, I'll read it, thanks!

beheh commented on src/config/index.tsx:
```diff
@@ -1,7 +1,12 @@
 
+ga("create", "UA-81509238-9", "auto");
+ga("set", "anonymizeIp", true);
+ga("send", "pageview", "config");
```
Let's make third parameter look like the real url:

```suggestion
ga("send", "pageview", "/config.html");
```

beheh commented on src/viewer/mobile/index.tsx:
```diff
@@ -4,13 +4,18 @@ import {
 
+ga("create", "UA-81509238-9", "auto");
+ga("set", "anonymizeIp", true);
+ga("send", "pageview", "mobile");
```
```suggestion
ga("send", "pageview", "/mobile.html");
```

beheh commented on src/viewer/overlay/index.tsx:
```diff
@@ -4,13 +4,18 @@ import {
 
+ga("create", "UA-81509238-9", "auto");
+ga("set", "anonymizeIp", true);
+ga("send", "pageview", "overlay");
```
```suggestion
ga("send", "pageview", "/overlay.html");
```

beheh commented on tsconfig.json:
```diff
@@ -5,7 +5,7 @@
     // "lib": [],                             /* Specify library files to be included in the compilation. */
     // "allowJs": true,                       /* Allow javascript files to be compiled. */
-    // "checkJs": true,                       /* Report errors in .js files. */
+    "checkJs": false,                         /* Report errors in .js files. */
```
Is this still even required?

edipo2s replied:
Tested again and it isn't.

beheh commented on src/viewer/Entity.tsx:
```diff
@@ -63,11 +63,22 @@ class Entity extends React.Component<Props & CardsProps & PortalProps, State> {
+			this.state.isMeaningfulHover
+		) {
+			if (this.props.dbfId) {
+				ga("send", "event", "Hover", "Card", String(this.props.dbfId));
```
```suggestion
				ga("send", "event", "Overlay", "Hover Card", String(this.props.dbfId));
```

beheh commented on src/viewer/Entity.tsx:
```diff
@@ -63,11 +63,22 @@ class Entity extends React.Component<Props & CardsProps & PortalProps, State> {
+			if (this.props.dbfId) {
+				ga("send", "event", "Hover", "Card", String(this.props.dbfId));
+			} else {
+				ga("send", "event", "Hover", "Card");
```
```suggestion
				ga("send", "event", "Overlay", "Hover Card");
```

beheh commented on src/viewer/overlay/DeckList.tsx:
```diff
@@ -350,6 +350,7 @@ class DeckList extends React.Component<
 											if (target && typeof target.blur === "function") {
 												target.blur();
 											}
+											ga("send", "event", "Click", "Copy Deck", "Overlay");
```
```suggestion
											ga("send", "event", "Deck", "Copy", "Overlay");
```

beheh commented on src/viewer/overlay/DeckList.tsx:
```diff
@@ -360,6 +361,7 @@ class DeckList extends React.Component<
 										<ShowButton
 											onClick={() => {
 												this.props.onPinned(false);
+												ga("send", "event", "Click", "Hide Deck");
```
```suggestion
												ga("send", "event", "Deck", "Hide");
```

beheh commented on src/viewer/overlay/DeckList.tsx:
```diff
@@ -370,6 +372,7 @@ class DeckList extends React.Component<
 										<HideButton
 											onClick={() => {
 												this.props.onPinned(true);
+												ga("send", "event", "Click", "Show Deck");
```
```suggestion
												ga("send", "event", "Deck", "Show");
```
